pkg/hakeeper/checkers/syshealth/check.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2021 - 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syshealth
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/hakeeper"
+	"github.com/matrixorigin/matrixone/pkg/hakeeper/checkers/util"
+	"github.com/matrixorigin/matrixone/pkg/hakeeper/operator"
+	hapb "github.com/matrixorigin/matrixone/pkg/pb/hakeeper"
+	"github.com/matrixorigin/matrixone/pkg/pb/logservice"
+)
+
+const (
+	defaultLogShardSize = 3
+)
+
+// Check checks system healthy or not.
+// If system wasn't healthy, we would generate
+// operators in order to shutdown all stores.
+func Check(
+	cluster hapb.ClusterInfo,
+	dnState hapb.DNState,
+	logState hapb.LogState,
+	currTick uint64,
+) ([]*operator.Operator, bool) {
+	sysHealthy := true
+
+	// parse all log stores for expired stores mainly
+	logStores := parseLogStores(logState, currTick)
+	if len(logStores.expired) == 0 {
+		return nil, sysHealthy
+	}
+
+	// check system healthy or not
+	for _, shard := range logShardsWithExpired(logStores.expired, logState, cluster) {
+		// if one of log shards wasn't healthy, the entire system wasn't healthy.
+		if !shard.healthy() {
+			sysHealthy = false
+			break
+		}
+	}
+
+	if sysHealthy {
+		return nil, sysHealthy
+	}
+
+	// parse all dn stores
+	dnStores := parseDnStores(dnState, currTick)
+
+	// generate operators to shutdown all stores
+	operators := make([]*operator.Operator, 0, logStores.length()+dnStores.length())
+	operators = append(operators, logStores.shutdownExpiredStores()...)
+	operators = append(operators, logStores.shutdownWorkingStores()...)
+	operators = append(operators, dnStores.shutdownExpiredStores()...)
+	operators = append(operators, dnStores.shutdownWorkingStores()...)
+
+	return operators, sysHealthy
```

Comment:
`sysHealthy` always returns `false` here.

Reply:
You should review code from line 38 to line 57 again.
In line 42, if there was no expired log store, the system was healthy, no problem.
In line 50, detailed in this PR, if there was a log shard with more than half replicas expired, then consider the entire system as unhealthy, we would stop all stores in the cluster.

Reply:
I mean you could return a constant `true` or `false` instead of a variable to make the code easy to understand.

Reply:
With a named variable, its intention in the logic is clear.
With a literal value, such as `true` or `false`, one must reference to its context in order to reason about its intention.
I appreciate your suggestion, but I don't think the original code is unclear in its meaning.